objects/isn_unifiedgrowingtray.lua

```diff
@@ -0,0 +1,524 @@
+		end
+	end
+	
+	--Compute which graphic should be displayed and update.
+	local growthperc = isn_getXPercentageOfY(storage.growth, storage.growthCap)
+	if growthperc >= 75 then
+		animator.setAnimationState("growth", "3")
+	elseif growthperc >= 50 then
+		animator.setAnimationState("growth", "2")
+	elseif growthperc >= 25 then
+		animator.setAnimationState("growth", "1")
+	else
+		animator.setAnimationState("growth", "0")
+	end
+	
+	--Check if a previous fluid use failed...
+	if storage.hasFluid == false then storage.hasFluid = isn_doFluidConsume() end
+	if storage.hasFluid ~= true then return end	--Failed again?  No growth nor outputs.
+	
+	storage.activeConsumption = true
+	
+	--Compute growth (we aren't dry right now)
+	storage.growth = storage.growth + ((storage.growthRate + storage.growthFluid + storage.growthFert) * growthmod)
+	
+	--Check if we should try to use more fluid due to growth.
+	if storage.currentStage < storage.stages then
+		--multiple stages might have happened, make sure we use apt fluid before resuming flow.
+		while true do
+			if storage.currentStage >= storage.stages then break end
+			if storage.growth < storage.stage[storage.currentStage].val then break end
+			if storage.hasFluid == false then break end
+			storage.hasFluid = isn_doFluidConsume()
+			storage.currentStage = storage.currentStage + 1
+		end
+	end
+	
+	--If the above fluid use left us dry stop here (don't output items) and try again later.
+	--Yes this may generate 1 more (unnecessary) growth spurt but it won't use more fluid.
+	if storage.hasFluid ~= true then return end
+	
+	--Check if we are grown.
+	if storage.growth >= storage.growthCap then
+		--Spawn our output...
+		for i = 1, storage.yield do
+			local itemDescriptors = nil
+			if storage.hasTree == true then
+				itemDescriptors = isn_genSaplingDrops()
+			elseif root.isTreasurePool(storage.harvestPool) then
+				itemDescriptors = root.createTreasure(storage.harvestPool, world.threatLevel())
+			end
+			if itemDescriptors ~= nil then
+				for _, itemDescriptor in pairs(itemDescriptors) do
+					local mask = {0, 1, 2}
+					if itemDescriptor.name == storage.currentseed.name then
+						mask = {1, 2}
+					end
+					fu_sendOrStoreItems(0, itemDescriptor, mask)
+					--next line might not be necessary... was in original growingtray but not hydroponicstray code.
+					--world.containerAddItems(entity.id(), itemDescriptor)
+				end
+			end
+		end
+		--Recycle the plant...
+		isn_doRecyclePlant()
+	end
+	
+end
+
+--Updates internal fluid levels and modifies a growth variable depending on
+--liquid in slot.
+function isn_doWaterIntake()
+	local contents = world.containerItems(entity.id())
+	local water = contents[storage.waterslot]
+	if water == nil then return false end
+	
+	for key, value in pairs(config.getParameter("isn_waterInputs")) do
+		if water.name == key and type(value) == "table" then
+			local amount, boost = value[1], value[2]
+			if amount == nil or boost == nil then
```

Comment:
`if not amount or not boost then`

Reply:
OK, this one is a bit fuzzy to me.  I get nil == false in the boolean context (ie type) but if the type is non-boolean?  If you know of a handy reference (searching has failed) I'd love to read it.

I'm far more comfortable with the explicit nil check in this context.

Is the following (pseudo-code) a valid way of thinking of things in this context?
Taking just amount for example, is this basically saying (conceptually)
```
branchcheck(variable)
  if variable == nil then return false end
  if type(variable) ~= boolean then return true end
  --if type(variable) == boolean then return valueof(variable) end
  return valueof(variable)
end
```

Also if that is valid then I'd expect the following to also work:
```
if not (amount or boost) then
```